### ISSUE TITLE
requires OS X 10.12 sierra?

### ISSUE THREAD

**Issue by willxy** [mbunkus/mkvtoolnix#1840]
On OS X 10.11.6, just downloaded 9.7.1. It's refusing to run, reporting:

> You can’t use this version of the application “MKVToolNix-9.7.1” with this version of OS X. You have OS X 10.11.6. The application requires OS X 10.12 or later.

**Comment by mbunkus** [comment]
I'll look into compiling for older versions.

**Comment by sagitt** [comment]
same

**Comment by mbunkus** [comment]
Another issue is that the 9.7.1 DMG is missing a library. Unless the user has Qt 5.7.1 installed this will lead to a crash when starting MKVToolNix GUI. I'm in the middle of fixing both issues.

**Comment by mbunkus** [comment]
Both issues should be fixed in [MKVToolNix-9.7.1-build2.dmg](https://mkvtoolnix.download/macos/MKVToolNix-9.7.1-build2.dmg). I'd appreciate some feedback.

**Comment by sagitt** [comment]
works on 10.12.2

**Comment by willxy** [comment]
The application gives me the same error ("The application requires OS X 10.12 or later.")  Interestingly, running the mkvtoolnix-gui from the cmdline seems to launch okay (I haven't tried muxing anything though).  I do notice the mkvinfo cmdline now runs (vs failing before about libQt5Concurrent).

**Comment by mbunkus** [comment]
I've tested the DMG I've linked in a Mavericks (10.9) installation and it starts up just fine. Did you perhaps run a non-fixed DMG by accident? Was the DMG still mounted? Was that a copy somewhere?

**Comment by willxy** [comment]
Strange.  Yes, I've redownloaded to be sure, and removed the prior dmg and .app; the .app is dated today.  If it helps, here's a screenshot of the finder Info, and the output I get from mdls of the .app.  You can see the icon overlay it has because it think it's not runnable.

![screen shot 2016-12-28 at 2 01 39 pm](https://cloud.githubusercontent.com/assets/18635433/21530482/b2cf67e6-cd06-11e6-808f-119678557af5.png)

```
kMDItemAlternateNames          = (
    "MKVToolNix-9.7.1.app"
)
kMDItemContentCreationDate     = 2016-12-28 18:47:56 +0000
kMDItemContentModificationDate = 2016-12-28 18:47:56 +0000
kMDItemContentType             = "com.apple.application-bundle"
kMDItemContentTypeTree         = (
    "com.apple.application-bundle",
    "com.apple.application",
    "public.executable",
    "com.apple.localizable-name-bundle",
    "com.apple.bundle",
    "public.directory",
    "public.item",
    "com.apple.package"
)
kMDItemDateAdded               = 2016-12-28 19:57:52 +0000
kMDItemDisplayName             = "MKVToolNix-9.7.1"
kMDItemExecutableArchitectures = (
    "x86_64"
)
kMDItemFSContentChangeDate     = 2016-12-28 18:47:56 +0000
kMDItemFSCreationDate          = 2016-12-28 18:47:56 +0000
kMDItemFSCreatorCode           = ""
kMDItemFSFinderFlags           = 0
kMDItemFSHasCustomIcon         = (null)
kMDItemFSInvisible             = 0
kMDItemFSIsExtensionHidden     = 1
kMDItemFSIsStationery          = (null)
kMDItemFSLabel                 = 0
kMDItemFSName                  = "MKVToolNix-9.7.1.app"
kMDItemFSNodeCount             = 1
kMDItemFSOwnerGroupID          = 20
kMDItemFSOwnerUserID           = 164
kMDItemFSSize                  = 70387945
kMDItemFSTypeCode              = ""
kMDItemKind                    = "Application"
kMDItemLanguages               = (
    English
)
kMDItemLogicalSize             = 70387945
kMDItemPhysicalSize            = 70803456
kMDItemVersion                 = "9.7.1"
```

**Comment by mbunkus** [comment]
Unfortunately that doesn't tell me anything; I'm not a macOS developer. All I can say is that it works fine for me on an even older macOS version than yours.

This is pretty much as far as I'm willing to go. If someone else has an idea what to do here, please speak up. I won't invest more time until then. Note that I'm currently building on Sierra with XCode 8.1 (Apple LLVM version 8.0.0 (clang-800.0.42.1). I've added `-mmacosx-version-min=10.9` during compilation of all libraries and applications included in the DMG.

The script I'm using for building everything (all the libraries, MKVToolNix and the DMG) is available in the repository in `tools/macos/build.sh`. If you (or anyone else) want to experiment with it and help me out, use a copy from git as it contains fixes not present in the release tarball. You'll have to adjust the paths in `tools/macos/config.sh` and download the source tarballs for the individual projects manually.

If that image isn't running for you, you should consider switching to [Homebrew](https://mkvtoolnix.download/downloads.html#macosx).

**Comment by willxy** [comment]
Thanks Moritz.  After downloading and prepping everything, I was able to build the dmg (on 10.11.6 with Xcode 8.2.1 Apple LLVM version 8.0.0 clang-800.0.42.1), and it starts up without complaint on 10.11.6.

Looking around, I found a few similar references to the error:

http://forum.startools.org/viewtopic.php?f=8&t=1098&start=0
http://community.openfl.org/t/mac-requires-os-x-10-12/8515
https://github.com/avibrazil/RDM/issues/5

The first indicated a similar oddity with running ok on 10.9.5 and complaining on 10.11.  The second referenced setting environment variable `MACOSX_DEPLOYMENT_TARGET=`, which 'net references suggest is synonymous with `-mmacosx-version-min=`:

http://stackoverflow.com/questions/25352389/difference-between-macosx-deployment-target-and-mmacosx-version-min-compiler-op

I was hoping my compile would produce a binary that still complained, so that I could test with `MACOSX_DEPLOYMENT_TARGET=` to see if it made a difference (eg, maybe there's somewhere in the chain that's not getting the cmdline argument).

So, I'm good to go with my own compile.  The only thing I could suggest is maybe adding `export MACOSX_DEPLOYMENT_TARGET=10.9` to the compile to see if that makes a difference even with the `-mmacosx-version-min=10.9`.  I don't have a machine running 10.12 yet to try it myself.

**Comment by mcringbearer** [comment]
In @mbunkus  build, it would appear the -mmacosx-version-min=10.9 flag is being ignored, because I was able to run the 9.7.1 build 2 on 10.8 and 10.9.   If that flag were really being observed, it shouldn't run on 10.8 but it does. The other reason I believe that flag is being ignored in @mbunkus build, is if you look at the info.plist file in the binary, the LSMinimumSystemVersion key is set to '10.8.0'.   For the record, the build2 ran on 10.8, 10.9, 10.12, but not 10.10 or 10.11.   

@willxy I think it would help Moritz tremendously if you would clearly spell out exactly what you did.  Sorry, but I read your post and the links you referenced and it is not clear to me what you did differently in your compile.  Is it simply <setenv name="MACOSX_DEPLOYMENT_TARGET" value="10.9"  that you set?   Would you be willing to post a link to your binary?

**Comment by willxy** [comment]
@mcringbearer Sorry for the unclarity.  I didn't end up doing anything different in my build; I used the build scripts as provided, and the resulting binary ran ok.  I'm currently assuming it ran ok on 10.11 simply because it was compiled on 10.11.

If it _hadn't_ run ok, I was going to try adding `export MACOSX_DEPLOYMENT_TARGET=10.9` to one of the bash scripts in `tools/macos/`.  I think testing that theory requires compiling on 10.12; I'm currently have only 10.11.

**Comment by mbunkus** [comment]
I appreciate the hints. Keep them coming, you're already sharing things I didn't know about :)

> @mcringbearer Sorry for the unclarity. I didn't end up doing anything different in my build; I used the build scripts as provided, and the resulting binary ran ok. I'm currently assuming it ran ok on 10.11 simply because it was compiled on 10.11.

I concur. I'm compiling on 10.12, and as far as I know clang defaults to target the OS version it is running on unless overwritten.

I currently have two macOS VMs that I'm using: an older Mavericks (10.9) one with which I built MKVToolNix up to and including 9.6.0, and a Sierra (10.12) one on which I built 9.7.x. The reason for the update is simply that the clang on Mavericks doesn't offer all the C++ features that MKVToolNix 9.7.0 and newer requires.

These are VMs running in VirtualBox on Windows. Setting them up was complicated and time-consuming enough; therefore I'm not really willing to try to set up one for 10.11. I'd much rather know why the resulting binary won't run on 10.11 even though I set the target to 10.9.

Note that during the compilation of Qt I've noticed that Qt's own build system sets `-mmacosx-version-min=10.8`, no matter what I add in `CFLAGS`/`CXXFLAGS` myself.

I'm willing to recompile everything with `MACOSX_DEPLOYMENT_TARGET` set; will try to fit that in tonight.

**Comment by mbunkus** [comment]
Another user has reported that he had to "unblock MKVToolNix.app from Gatekeeper in the system preferences" as I don't code-sign the DMG yet. Maybe that'd explain why my DMG doesn't run for you?

**Comment by willxy** [comment]
Gatekeeper -- that's not it, no; mine is already set to "allow apps downloaded from: anywhere" (vs from: "Mac App Store", or from: "Mac App Store and identified developers").

**Comment by Marc66** [comment]
I confirm willxy's post. Same thing here (allow all Apps in Gatekeeper) and new binary still does not work under OSX 10.11.6

**Comment by mbunkus** [comment]
I've created two new DMGs, both compiled on Sierra, one with [`MACOSX_DEPLOYMENT_TARGET=10.9`](https://mkvtoolnix.download/macos/pre/MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.9.dmg), one with [`MACOSX_DEPLOYMENT_TARGET=10.10`](https://mkvtoolnix.download/macos/pre/MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.10.dmg). I also tried to build with `MACOSX_DEPLOYMENT_TARGET=10.8`, but boost wouldn't compile under those settings.

As my Mavericks VM doesn't boot anymore I cannot even test them myself, unfortunately.

**Comment by mcringbearer** [comment]
Very interesting!  The build by @willxy runs on 10.8 (Mountain Lion), 10.9 and 10.11, but NOT 10.10.  Results in the "You have OS X 10.10.5. The application requires OS X 10.11 or later."
So it would appear that on 10.8 and 10.9,  both the @willxy build and the @mbunkus build 2 both run (apparently the minimum flag is simply IGNORED).  Otherwise, it seems the system used to create the build defines the minimum OS required, regardless of any flags or env variables.

**Comment by mcringbearer** [comment]
@mbunkus if you provide links to the two new DMGs you created, I can test them.  I have VMs for 10.8, 10.9, 10.10 and 10.11 (setup under VMWare Fusion 8) I can quickly test with.  I'm not a 'developer' per se, I just dabble, strictly amatuer, but I suspect the issues here may be QT.   I recall the developer of Calibre, which is also multi platform, Qt based, had a similar problem when 10.10 came out and he updated to Qt 5.  The developer did eventually resolve it, don't know how.  I found this link/discussion which points to a Qt compiler option which you did not mention, QMAKE_MACOSX_DEPLOYMENT_TARGET (note the 'QMAKE' in front), located in the Qmake config file for CLANG.  This is not an env variable, but actual line in config file for Qt.  MIght be a clue; wish I was more competent to help.  Here's the article:
http://stackoverflow.com/questions/24243176/how-to-specify-target-mac-os-x-version-using-qmake

I found this link by searching in google for 'os x qt5 minimum system' and there are many other links that came up discussing similar problems, i.e. targeting a minimum OS X version.

**Comment by willxy** [comment]
@mcringbearer The 10.9 and 10.10 dmg's are linked in his comment above (they show up as blue for me, as clickable links).

@mbunkus Both dmg's launch successfully for me under 10.11, ie both the `MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.9.dmg` and the `MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.10.dmg`.  Looks like the `MACOSX_DEPLOYMENT_TARGET` may do the trick.

**Comment by mcringbearer** [comment]
Sooo... the two builds by @mbunkus, using the TARGET-10.9 and TARGET-10.10 flags, both CRASH on 10.8, 10.9, 10.10 and 10.11.  These were run on systems without Xcode or Qt developer tools installed, which is what most users will have.  The crash is the same, the app runs the UpdateCheckThread (presumably to check for newer version of MKVToolNix?), it trys to make a call to _clock_gettime in the libSystem.B.dylib file.  Here's the first line of the crash report:

Dyld Error Message:
  Symbol not found: _clock_gettime
  Referenced from: /Users/USER/Desktop/MKVToolNix-9.7.1.app/Contents/MacOS/mkvtoolnix-gui
  Expected in: /usr/lib/libSystem.B.dylib

Did some Google searching on "libSystem.B.dylib crash _clock_gettime" and apparently this is a known issue with XCode 8 and 10.11.6 (and presumbably 10.12 as well, since @mbunkus compiled this on 10.12.)   Reading some of the links, app compiles fine but fails at runtime.     The possible reason it runs on @willxy's 10.11 system is because you have Xcode and QT dev installed, so the call is successfully resolved.   It runs fine on @mbunkus Sierra system and other people with Sierra because according to one of the links, Apple introduced _clock_gettime in Sierra.  A fix is described here:
https://github.com/pocoproject/poco/issues/1453 
it involves checking to see if the symbol for _clock_gettime is defined at compile time and taking different action depending on the result.

The 'simplest' solution at this point, barring trying to figure out what effect all these flags have, is to compile on a system using the earliest version of OS X that still satisfies all the other Qt and C++ libraries requirements that MKVToolnix requires.  Whatever that version is (10.10?) appears to dictate what minimum version the app will run on.  But in any case, totally separate issue, compiling on Sierra and changing MACOSX_DEPLOYMENT_TARGET= to anything prior to Sierra, will result in the _get_clocktime error for anyone not running Sierra.

Right now, the 'build2' option is still the most functional, as you only get 'minimum system' error, whereas messing with MACOSX_DEPLOYMENT_TARGET will result in a crash if you don't have Xcode installed.

**Comment by willxy** [comment]
Hmm, oddly, what I observed just now was the same crash (_clock_gettime), where earlier I was able to launch it, and run a merge.  I'm not sure what changed between then and now.  I'll keep looking.

**Comment by mbunkus** [comment]
> Symbol not found: _clock_gettime

Oh right, that `_clock_gettime` issue. I've run into it, too, and "solved" it by using an older version of libcurl that I had compiled back on Mavericks. That's how the released `build2` DMG was built.

> Hmm, oddly, what I observed just now was the same crash (_clock_gettime), where earlier I was able to launch it, and run a merge.

The fact that you're only observing the crash now is perfectly explainable. MKVToolNix GUI only attempts one update check per 24 hours. Due to your experiments with other DMGs you were likely still within that cooldown period.

I'll try to build two new DMGs with `MACOSX_DEPLOYMENT_TARGET` set appropriately and with libcurl from Mavericks later tonight.

> I found this link/discussion which points to a Qt compiler option which you did not mention, `QMAKE_MACOSX_DEPLOYMENT_TARGET` (note the 'QMAKE' in front), located in the Qmake config file for CLANG.

I'm not using a qmake-based build system, therefore this setting doesn't apply to me. I guess that setting `QMAKE_MACOSX_DEPLOYMENT_TARGET` in qmake's project configuration causes qmake to set the environment variable `MACOSX_DEPLOYMENT_TARGET` or to add the compiler option `-mmacosx-version-min`.

**Comment by mbunkus** [comment]
Here are two DMGs with older libcurls:

  * [MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.9-curl-from-mavericks.dmg](https://mkvtoolnix.download/macos/pre/MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.9-curl-from-mavericks.dmg)
  * [MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.10-curl-from-mavericks.dmg](https://mkvtoolnix.download/macos/pre/MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.10-curl-from-mavericks.dmg)

As before untested on machines earlier than Sierra due to Mavericks not starting anymore. I'd appreciate some feedback.

**Comment by mbunkus** [comment]
As a side note. I've just pushed a couple of commits that rework network access to use Qt's classes instead of CURL. The next release of MKVToolNix won't use CURL anymore.

**Comment by mcringbearer** [comment]
So here's the GREAT news, both Builds with the older libcurls run on 10.8, 10.9, 10.10 and 10.11.  (I didn't test Sierra 10.12, as I don't have a VM of that yet).  The check for new version at startup occurs without issue (I use Little Snitch, and it tells me the app is requesting connection to mkvtoolnix/download).  So the _clock_gettime is fixed.

Sadly though, something isn't quite right, as in theory, if the flags were working as expected, NEITHER build should work on 10.8, but BOTH do, and the one build with the 10.10 flag should NOT run on 10.9 either, but it does.   Looking at the info.plist file in both builds, the 'LSMinimumSystemVersion' key is set to '10.8.0' which I guess explains why both can run on 10.8 and higher, so those flags do result in a build that runs, but not the way you'd expect, as in theory, it should change the 'LSMinimumSystemVersion' key.

Perhaps this is a case where we 'leave well enough alone' as they say.  Thank you mbunkus for working this through, on behalf of the users who can now still run on 10.8 and higher, we appreciate it and thank you.  Looking forward to a new release build with the fix.  In the absence of any other input, I suppose you could pick either build to release, but if it were me I'd probably go with the one where the target is set to 10.9, just because it's an earlier release in case there is anything else we don't know that still might be version specific that is actually looking at that flag.

**Comment by mbunkus** [comment]
> So here's the GREAT news, both Builds with the older libcurls run on 10.8, 10.9, 10.10 and 10.11. (I didn't test Sierra 10.12, as I don't have a VM of that yet). The check for new version at startup occurs without issue (I use Little Snitch, and it tells me the app is requesting connection to mkvtoolnix/download). So the _clock_gettime is fixed.

`_clock_gettime` wasn't really the big problem here — macOS refusing to start the application was. But it's good to hear that the builds do work. Thanks for the feedback.

> Looking at the info.plist file in both builds, the 'LSMinimumSystemVersion' key is set to '10.8.0'

This is due to [my build script always writing 10.8.0](https://github.com/mbunkus/mkvtoolnix/blob/8eec6faf0c1bcc0dd14f82dcfdf6409e7aabec65/tools/macos/build.sh#L313) when creating the DMG, I guess. I think I'll change the build script to write the same property that's used in `MACOSX_DEPLOYMENT_TARGET`.

> I suppose you could pick either build to release, but if it were me I'd probably go with the one where the target is set to 10.9, just because it's an earlier release in case there is anything else we don't know that still might be version specific that is actually looking at that flag.

Yeah. I'll do one more build with `MACOSX_DEPLOYMENT_TARGET=10.9` and the `LSMinimumSystemVersion` property set to match and upload it for you guys to test. If that works, I'll upload that DMG as the new official 9.7.1 build.

**Comment by mbunkus** [comment]
Please test this build: [MKVToolNix-9.7.1-build3.dmg](https://mkvtoolnix.download/macos/pre/MKVToolNix-9.7.1-build3.dmg). Thanks.

**Comment by willxy** [comment]
On my 10.11:

- `MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.9-curl-from-mavericks.dmg`
- `MKVToolNix-9.7.1-host-10.12-MACOSX_DEPLOYMENT_TARGET-10.10-curl-from-mavericks.dmg` 

launch okay, and:

- `MKVToolNix-9.7.1-build3.dmg`

seems to be giving me the `_clock_gettime` crash.

**Comment by mbunkus** [comment]
Meh… Seems I accidentally built curl again. I'll throw everything away & rebuild everything but curl. Expect a new DMG in an hour.

**Comment by mbunkus** [comment]
And here's [MKVToolNix-9.7.1-build4.dmg](https://mkvtoolnix.download/macos/pre/MKVToolNix-9.7.1-build4.dmg).

**Comment by willxy** [comment]
Awesome, build4 launched just fine for me.

**Comment by mcringbearer** [comment]
Will teest build 4 this afternoon and get back to y'all....sorry can't get to it any sooner :-(

**Comment by mcringbearer** [comment]
Build 4 runs on 10.9, 10.10 and 10.11.  Fails on 10.8 as expected with the usual "This app requires 10.9 or later to run....".  Thanks mbunkus!   PS. on the Macupdate/mkvtoolnix site, I'm the user with the avatar 'MCR' that has posted comments.   "SHIP IT!" as we used to say in the old days when software was still physically distributed.

**Comment by mbunkus** [comment]
Done. Thanks for all the feedback & the tests.

**Comment by willxy** [comment]
Awesome.  Thanks Moritz!

**Comment by mcringbearer** [comment]
Moritz I posted a comment on MacUpdate advising people of the availability of build#4.  Peace.
